### PR TITLE
Refactor tests

### PR DESCRIPTION
**Pull request by sequencer**
It deprecate `chisel3.testers`, which should not be a user API.
API `TreadleBackend` in `tests` package is not visible to `chisel3` package, so I move all APIs to `chisel3Tests` package, and refactor the imports.
After adding 7687c1d
sbt will add the compiler plugin automatically from chisel project???
Tuned this for the whole weekend, and gave up.

### Contributor Checklist

- [ ] Did you add Scaladoc to every public function/method?
- [ ] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [ ] Did you add appropriate documentation in `docs/src`?
- [x] Did you state the API impact?
- [x] Did you specify the code generation impact?
- [x] Did you request a desired merge strategy?
- [ ] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement
- code refactoring

#### API Impact
deprecate APIs

#### Backend Code Generation Impact
None

#### Desired Merge Strategy
- Rebase: You will rebase the PR onto master and it will be merged with a merge commit.

#### Release Notes
None

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels?
- [ ] Did you mark the proper milestone (3.2.x, 3.3.x, 3.4.x, 3.5.0) ?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you mark as `Please Merge`?

### CODE REVIEW

**Comment by ekiwi** [comment]
Could you investigate whether there is a way to enable/disable certain tests depending on whether the action was triggered by a push or a PR? I would prefer to avoid having two completely separate github action files since that increases the maintenance burden.

**Comment by sequencer** [comment]
> Could you investigate whether there is a way to enable/disable certain tests depending on whether the action was triggered by a push or a PR? I would prefer to avoid having two completely separate github action files since that increases the maintenance burden.

Done:)

**Comment by sequencer** [comment]
I'm not familiar with sbt,
I'm finding a way to make `no-plugin-test` depends on `chisel3 / Test` while not using the plugin.
sbt sucks...

**Comment by sequencer** [comment]
OMG never image we can use symlink to do this hack! Thanks @jackkoenig

File: src/main/scala/chisel3/testers/BasicTester.scala [hunk 1]
```diff
@@ -9,6 +9,7 @@ import chisel3.internal.Builder.pushCommand
 import chisel3.internal.firrtl._
 import chisel3.internal.sourceinfo.SourceInfo
 
+@deprecated("This is not a User API, will be removed in Chisel 3.6", "3.5")
 class BasicTester extends Module() {
```
Comment (sequencer) [inline src/main/scala/chisel3/testers/BasicTester.scala:13]: Deprecation here.

File: src/main/scala/chisel3/testers/TesterDriver.scala [hunk 2]
```diff
@@ -13,6 +13,7 @@ import firrtl.options.{Dependency, Phase, PhaseManager, TargetDirAnnotation, Uns
 import firrtl.stage.{FirrtlCircuitAnnotation, FirrtlStage}
 import firrtl.transforms.BlackBoxSourceHelper.writeResourceToDirectory
 
+@deprecated("This is not a User API, will be removed in Chisel 3.6", "3.5")
 object TesterDriver extends BackendCompilationUtilities {
```
Comment (sequencer) [inline src/main/scala/chisel3/testers/TesterDriver.scala:17]: Deprecation here.

**Comment by sequencer** [comment]
subsumed by #2043

**Pull request closed without merging**